fem/bilininteg_ext.cpp

```diff
@@ -0,0 +1,1581 @@
+            {
+               gradXY[dy][dx][0] = 0;
+               gradXY[dy][dx][1] = 0;
+               gradXY[dy][dx][2] = 0;
+            }
+         }
+         for (int qy = 0; qy < Q1D; ++qy)
+         {
+            double gradX[D1D][4];
+            for (int dx = 0; dx < D1D; ++dx)
+            {
+               gradX[dx][0] = 0;
+               gradX[dx][1] = 0;
+               gradX[dx][2] = 0;
+            }
+            for (int qx = 0; qx < Q1D; ++qx)
+            {
+               const double gX = grad[qz][qy][qx][0];
+               const double gY = grad[qz][qy][qx][1];
+               const double gZ = grad[qz][qy][qx][2];
+               for (int dx = 0; dx < D1D; ++dx)
+               {
+                  const double wx  = Bt(dx,qx);
+                  const double wDx = Gt(dx,qx);
+                  gradX[dx][0] += gX * wDx;
+                  gradX[dx][1] += gY * wx;
+                  gradX[dx][2] += gZ * wx;
+               }
+            }
+            for (int dy = 0; dy < D1D; ++dy)
+            {
+               const double wy  = Bt(dy,qy);
+               const double wDy = Gt(dy,qy);
+               for (int dx = 0; dx < D1D; ++dx)
+               {
+                  gradXY[dy][dx][0] += gradX[dx][0] * wy;
+                  gradXY[dy][dx][1] += gradX[dx][1] * wDy;
+                  gradXY[dy][dx][2] += gradX[dx][2] * wy;
+               }
+            }
+         }
+         for (int dz = 0; dz < D1D; ++dz)
+         {
+            const double wz  = Bt(dz,qz);
+            const double wDz = Gt(dz,qz);
+            for (int dy = 0; dy < D1D; ++dy)
+            {
+               for (int dx = 0; dx < D1D; ++dx)
+               {
+                  y(dx,dy,dz,e) +=
+                     ((gradXY[dy][dx][0] * wz) +
+                      (gradXY[dy][dx][1] * wz) +
+                      (gradXY[dy][dx][2] * wDz));
+               }
+            }
+         }
+      }
+   });
+}
+
+typedef void (*fDiffusionMultAdd)(const int NE,
+                                  const double* B,
+                                  const double* G,
+                                  const double* Bt,
+                                  const double* Gt,
+                                  const double* oper,
+                                  const double* solIn,
+                                  double* solOut);
+
+static void PADiffusionMultAssembled(const int dim,
+                                     const int D1D,
+                                     const int Q1D,
+                                     const int NE,
+                                     const double* B,
+                                     const double* G,
+                                     const double* Bt,
+                                     const double* Gt,
+                                     const double* op,
+                                     const double* x,
+                                     double* y)
+{
```

Comment:
Will the ifdef for libceed (and future implementations) come on this level?

Reply:
Yes, see the okina+libceed PR, https://github.com/mfem/mfem/pull/804

Reply:
As these functions `(PADiffusionMultAssembled`, `PADiffusionAssemble`, `PAMassMultAssembled` ...) are the fork between different kernel implementations, any addition of a new kernel module, like libceed, will require going through all and changing them. Also adding a new integrator will require calling methods from all kernel modules.

Maybe we can have another layer in between, so that these functions make a call like `Device::GetActiveDeviceKernels()->DiffusionMultAssembledKernel(...)`. Here `DeviceKernels` is abstract class with methods for all kernels, corresponding to the different `BilinearFormIntegrators`. As the `MFEM_FORALL` does its own switch inside, this can be in the `OkinaDeviceKernels`.

Then when libceed comes, all new code will be in `LibCEEDDeviceKernels`, with no changes here.

Reply:
Did you check the branch #804  to see if what you propose would be doable? I'm planning to do some cleaning but this is my first proposal for the integration of libCEED.